### ISSUE TITLE
Carpenter fails when bins with custom integer edges are set

### ISSUE THREAD

**Issue by benkrikler** [benkrikler/fast-carpenter-github-test#16]
**Imported from [gitlab issue 16](https://gitlab.cern.ch/fast-hep/public/fast-carpenter/issues/16)**

**Steps to reproduce:**
1) Define a ```fast_carpenter.summary.BinnedDataframe```
2) Define integer custom bin edges, e.g.:
```yaml
MyBinner:
  dataset_col: True
  binning:
          - {in: genJet_pt, out: genJetPt, bins: {edges:[25, 30]}}
```
3) Run fast_carpenter

The carpenter fails with
```bash
Traceback (most recent call last):
  File "/users/sb17498/.local/bin/fast_carpenter", line 11, in <module>
    sys.exit(main())
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_carpenter/__main__.py", line 51, in main
    sequence = fast_flow.read_sequence_yaml(args.sequence_cfg, output_dir=args.outdir)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_flow/v1/__init__.py", line 19, in read_sequence_yaml
    return sequence_from_dict(**cfg)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_flow/v1/dict_config.py", line 28, in sequence_from_dict
    stages = _configure_stages(stages, output_dir, stage_descriptions)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_flow/v1/dict_config.py", line 50, in _configure_stages
    stage = stage_class(**cfg)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_carpenter/summary.py", line 52, in __init__
    ins, outs, binnings = _create_binning_list(self.name, binning)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_carpenter/summary.py", line 156, in _create_binning_list
    _in, _out, _bins, _index = _create_one_dimension(name, **cleaned_dimension_dict)
  File "/users/sb17498/.local/lib/python2.7/site-packages/fast_carpenter/summary.py", line 192, in _create_one_dimension
    bin_obj = np.insert(bin_obj, 0, float("-inf"))
  File "/users/sb17498/.local/lib/python2.7/site-packages/numpy/lib/function_base.py", line 5061, in insert
    values = array(values, copy=False, ndmin=arr.ndim, dtype=arr.dtype)
OverflowError: cannot convert float infinity to integer
```

What is going on is that numpy sees a list of integer and defines a integer array. Later when underflow and overflow edges are implemented, ```bin_obj = np.insert(bin_obj, 0, float("-inf"))``` forces a float inf value to be put in a integer array, causing the crash.

**Temporary workaround:** Define at least one edge as ```float```, e.g. 25. or 25.0, that sets the entire array as float and fixes the issue.

I presume the actual solution is forcing the array to be always ```float```.

**Comment by benkrikler** [comment]
**On 2018-11-23 Simone Bologna (sbologna) wrote:**

changed the description

**Comment by benkrikler** [comment]
**On 2018-11-27 Benjamin Krikler (bkrikler) wrote:**

@sbologna thanks for flagging this.  It's great to have another set of eyes on the code.  The fix you suggest sounds reasonable to me; ideally we'd come up with something else which allows integer bins to remain integers but forcing things to floats sounds like the simplest at this point.  Would you be up for trying to implement this yourself and then opening an MR?

**Comment by benkrikler** [comment]
**On 2018-11-27 Simone Bologna (sbologna) wrote:**

Hi,

I think I can, it does not look hard to do.

I'll keep you posted.

Simone

**Comment by benkrikler** [comment]
**On 2018-11-27 Simone Bologna (sbologna) wrote:**

mentioned in merge request [gitlab:!21](https://gitlab.cern.ch/fast-hep/public/fast-carpenter/merge_requests/21)

**Comment by benkrikler** [comment]
**On 2018-11-27 Benjamin Krikler (bkrikler) wrote:**

closed via commit 73a4334334c95255fb7d8e704a97eb540ae82d55

**Comment by benkrikler** [comment]
**On 2018-11-27 Benjamin Krikler (bkrikler) wrote:**

closed via commit 8c5b9a3bc029cfa68dd76aeb997109db12b56ad5
